File: generate_dataset.js
const createCsvWriter = require('csv-writer').createObjectCsvWriter;
const crypto = require('crypto');

const csvWriter = createCsvWriter({
    path: 'problems.csv',
    header: [
        {id: 'title', title: 'title'},
        {id: 'description', title: 'description'},
        {id: 'input_description', title: 'input_description'},
        {id: 'output_description', title: 'output_description'},
        {id: 'problem_class', title: 'problem_class'},
        {id: 'problem_score', title: 'problem_score'}
    ]
});

const easyKeywords = ["sum", "array", "numbers", "add", "minimum", "maximum", "simple", "integer", "print", "count"];
const mediumKeywords = ["sorting", "binary search", "greedy", "dynamic programming", "graph", "BFS", "DFS", "tree", "matrix", "optimization", "recursion"];
const hardKeywords = ["max flow", "segment tree", "heavy-light decomposition", "FFT", "geometry", "persistent data structure", "bitmask DP", "Centroid Decomposition", "flow", "suffix automaton"];

const mathSymbols = ["+", "-", "*", "/", ">", "<", "=", "∑", "∏", "log", "exp", "sqrt", "%"];

function getRandomItem(array) {
    return array[Math.floor(Math.random() * array.length)];
}

function generateRandomDescription(category) {
    let keywords, length, symbolsCount;
    
    if (category === "Easy") {
        keywords = Array.from({length: 3}, () => getRandomItem(easyKeywords));
        length = 20 + Math.floor(Math.random() * 20);
        symbolsCount = 1 + Math.floor(Math.random() * 3);
    } else if (category === "Medium") {
        keywords = Array.from({length: 5}, () => getRandomItem([...easyKeywords, ...mediumKeywords]));
        length = 40 + Math.floor(Math.random() * 40);
        symbolsCount = 3 + Math.floor(Math.random() * 5);
    } else { // Hard
        keywords = Array.from({length: 8}, () => getRandomItem([...mediumKeywords, ...hardKeywords]));
        length = 80 + Math.floor(Math.random() * 80);
        symbolsCount = 8 + Math.floor(Math.random() * 10);
    }
    
    const words = ["In", "this", "problem", "you", "need", "to", "calculate", "the", "given", "value", "using", "the", "algorithm", "and", "data", "structures", "provided", "in", "the", "input", "Each", "case", "contains", "multiple", "integers", "representing", "the", "constraints", "and", "requirements"];
    
    let textArray = Array.from({length}, () => getRandomItem(words));
    
    // Inject keywords
    keywords.forEach(kw => {
        textArray[Math.floor(Math.random() * textArray.length)] = kw;
    });
    
    // Inject math symbols
    for (let i = 0; i < symbolsCount; i++) {
        textArray[Math.floor(Math.random() * textArray.length)] = getRandomItem(mathSymbols);
    }
    
    return textArray.join(" ");
}

const records = [];
for (let i = 0; i < 1000; i++) {
    const classes = ["Easy", "Medium", "Hard"];
    const weights = [0.4, 0.4, 0.2];
    const rand = Math.random();
    let problem_class = "Easy";
    if (rand < 0.4) problem_class = "Easy";
    else if (rand < 0.8) problem_class = "Medium";
    else problem_class = "Hard";
    
    let problem_score;
    if (problem_class === "Easy") problem_score = 800 + Math.floor(Math.random() * 400);
    else if (problem_class === "Medium") problem_score = 1300 + Math.floor(Math.random() * 700);
    else problem_score = 2000 + Math.floor(Math.random() * 1500);
    
    records.push({
        title: `Problem ${1000 + i}`,
        description: generateRandomDescription(problem_class),
        input_description: generateRandomDescription(problem_class),
        output_description: generateRandomDescription(problem_class),
        problem_class: problem_class,
        problem_score: problem_score
    });
}

csvWriter.writeRecords(records)
    .then(() => console.log('Generated problems.csv with 1000 samples.'));

File: generate_dataset.py
import pandas as pd
import numpy as np
import random

# Seed for reproducibility
random.seed(42)
np.random.seed(42)

def generate_random_description(category):
    easy_keywords = ["sum", "array", "numbers", "add", "minimum", "maximum", "simple", "integer", "print"]
    medium_keywords = ["sorting", "binary search", "greedy", "dynamic programming", "graph", "BFS", "DFS", "tree", "matrix", "optimization"]
    hard_keywords = ["max flow", "segment tree", "heavy-light decomposition", "FFT", "geometry", "persistent data structure", "bitmask DP", "Centroid Decomposition"]
    
    math_symbols = ["+", "-", "*", "/", ">", "<", "=", "∑", "∏", "log", "exp"]
    
    if category == "Easy":
        keywords = random.sample(easy_keywords, k=random.randint(2, 4))
        length = random.randint(15, 30)
        symbols_count = random.randint(1, 3)
    elif category == "Medium":
        keywords = random.sample(medium_keywords + easy_keywords, k=random.randint(3, 6))
        length = random.randint(30, 60)
        symbols_count = random.randint(3, 7)
    else: # Hard
        keywords = random.sample(hard_keywords + medium_keywords, k=random.randint(5, 8))
        length = random.randint(60, 120)
        symbols_count = random.randint(8, 15)
        
    # Generate fake problem text
    words = ["Lorem", "ipsum", "dolor", "sit", "amet", "consectetur", "adipiscing", "elit", "sed", "do", "eiusmod", "tempor"]
    text = " ".join(random.choices(words, k=length))
    
    # Inject keywords and math symbols
    text_list = text.split()
    for kw in keywords:
        text_list[random.randint(0, len(text_list)-1)] = kw
    for _ in range(symbols_count):
        text_list[random.randint(0, len(text_list)-1)] = random.choice(math_symbols)
        
    return " ".join(text_list)

def generate_dataset(n=500):
    data = []
    for _ in range(n):
        problem_class = random.choices(["Easy", "Medium", "Hard"], weights=[0.4, 0.4, 0.2])[0]
        
        if problem_class == "Easy":
            score = random.randint(800, 1200)
        elif problem_class == "Medium":
            score = random.randint(1300, 1900)
        else:
            score = random.randint(2000, 3500)
            
        title = f"Problem {random.randint(1000, 9999)}"
        description = generate_random_description(problem_class)
        input_desc = generate_random_description(problem_class)
        output_desc = generate_random_description(problem_class)
        
        data.append({
            "title": title,
            "description": description,
            "input_description": input_desc,
            "output_description": output_desc,
            "problem_class": problem_class,
            "problem_score": score
        })
        
    df = pd.DataFrame(data)
    df.to_csv("problems.csv", index=False)
    print(f"Generated dataset with {n} samples saved to problems.csv")

if __name__ == "__main__":
    generate_dataset(1000)
